### PR TITLE
 use of Result::unwrap_or_else over Result::map_or_else

### PR DESCRIPTION
**Pull request by Valentine-Mario**
This PR is the lint suggested in #7328

### CODE REVIEW

**Comment by llogiq** [comment]
Hey there! I see you have put the basic structure together, now you may want to start fleshing out the test, then implement the actual lint.

Let me know if you need any help.

**Comment by Valentine-Mario** [comment]
> Hey there! I see you have put the basic structure together, now you may want to start fleshing out the test, then implement the actual lint.
> 
> Let me know if you need any help.

Thank you. I definitely would reach out when I need help

**Comment by Valentine-Mario** [comment]
> Hey there! I see you have put the basic structure together, now you may want to start fleshing out the test, then implement the actual lint.
> 
> Let me know if you need any help.

Hey, @llogiq I'm currently facing a challenge. So my test is passing even when they are not supposed to, thus I can't run `cargo dev bless`. But I've written a simple linter that would check for map_or_else method ident and verify if it is a closure. Your help would be greatly appreciated. Thanks

**Comment by Valentine-Mario** [comment]
@llogiq  can you please help me with review

File: clippy_lints/src/unwrap_or_else_over_map_or_else.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    ///    Ok("An even number :)")
+    /// }
+    ///
+    /// func_result(2).map_or_else(|e| println!("{:?}", e), |n| println!("{}", n))
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    ///  /// let c= func_result(3).unwrap_or_else(|e| {
+    ///     println!("Error: {:?}", e);
+    ///    process::exit(1);
```
Comment (llogiq) [inline clippy_lints/src/unwrap_or_else_over_map_or_else.rs:32]: I somehow doubt that exiting the process makes for a good suggestion.

**Review by llogiq — COMMENTED** [review]
I feel the example suggestion is worse than the code it replaces. Ultimately I don't understand the reasoning behind this lint.

Both `map_or_else` and `unwrap_or_else` are about getting the return value, yet your examples only ever return `()`. Worse, `process::exit` for control flow is not generally applicable and in most cases misguided.

**Comment by llogiq** [comment]
Having read the underlying issue, I think the lint was meant to catch an `id`-implementing closure in the `map` part of the `map_or_else`.

**Comment by Valentine-Mario** [comment]
> I feel the example suggestion is worse than the code it replaces. Ultimately I don't understand the reasoning behind this lint.
> 
> Both `map_or_else` and `unwrap_or_else` are about getting the return value, yet your examples only ever return `()`. Worse, `process::exit` for control flow is not generally applicable and in most cases misguided.

I've updated the doc and test to give a clearer picture to the lint usage

**Comment by llogiq** [comment]
I'm still not happy about the example.

Things I would consider lint-worthy:

`a.map_or_else(|_| 0, |n| n)` (but only in the absence of type adjustments) or `b.map_or_else(|x| something_else(x), std::convert::identity)`.

Note that both are expressions that return some value (of the `Ok`-part of the type of `a` and `b`, respectively).

As soon as the `map` part does some operation, I would consider linting the expression a false positive.

**Comment by Valentine-Mario** [comment]
> I'm still not happy about the example.
> 
> Things I would consider lint-worthy:
> 
> `a.map_or_else(|_| 0, |n| n)` (but only in the absence of type adjustments) or `b.map_or_else(|x| something_else(x), std::convert::identity)`.
> 
> Note that both are expressions that return some value (of the `Ok`-part of the type of `a` and `b`, respectively).
> 
> As soon as the `map` part does some operation, I would consider linting the expression a false positive.

Thanks for that. I'm kind of new to contributing, so how do I check if the map part performs any form of operation?

**Comment by llogiq** [comment]
For that you need to match on the second argument expression of the `map_or_else` method call (`args[2]`, because `[0]` is the receiver). Either it is an `ExprPath`, then you check if that matches `std::convert::identity`. Or it is an `ExprClosure` then you need to get the `Body` and see if it is an `ExprPath` with just one segment and the same ident as the closure's sole argument. In this case, check no type adjustments took place (see `eta_reduction.rs` as a reference).

**Comment by flip1995** [comment]
> how do I check if the map part performs any form of operation?

With #7330 you can do this with the function `clippy_utils::is_expr_identity_function`

**Comment by camsteffen** [comment]
Any thoughts on enhancing `map_identity` instead of creating a new lint? We could also include `Result::map_or`, `Option::map_or`, `Option::map_or_else`.

**Comment by Valentine-Mario** [comment]
> For that you need to match on the second argument expression of the `map_or_else` method call (`args[2]`, because `[0]` is the receiver). Either it is an `ExprPath`, then you check if that matches `std::convert::identity`. Or it is an `ExprClosure` then you need to get the `Body` and see if it is an `ExprPath` with just one segment and the same ident as the closure's sole argument. In this case, check no type adjustments took place (see `eta_reduction.rs` as a reference).

Thanks for the explanation. I used a slightly different approach though. After getting the closure expr body, I checked if it was an expr block. Then if  block.stmts.is_empty(), then we could say that there is no type adjustments

**Comment by Valentine-Mario** [comment]
> For that you need to match on the second argument expression of the `map_or_else` method call (`args[2]`, because `[0]` is the receiver). Either it is an `ExprPath`, then you check if that matches `std::convert::identity`. Or it is an `ExprClosure` then you need to get the `Body` and see if it is an `ExprPath` with just one segment and the same ident as the closure's sole argument. In this case, check no type adjustments took place (see `eta_reduction.rs` as a reference).

Thanks for this. In my case, I checked if it was an ```ExprKind::Lit```, then I checked the adjstments

**Comment by llogiq** [comment]
As @flip1995 said, we already have a utils function to find identities. You can simply use that one.

And @camsteffen is right, we should extend existing lints that fit before creating a new one.

**Comment by Valentine-Mario** [comment]
> As @flip1995 said, we already have a utils function to find identities. You can simply use that one.
> 
> And @camsteffen is right, we should extend existing lints that fit before creating a new one.

Not sure how the extending should work. Since map_identity matches for just `map` method name

**Comment by camsteffen** [comment]
> Not sure how the extending should work. Since map_identity matches for just `map` method name

You can just edit the implementation of `map_identity` to match `map_or_else`. Ultimately you just need to do `span_lint_and_help(cx, MAP_IDENTITY, ..)`.

**Comment by giraffate** [comment]
ping from triage @Valentine-Mario. Can you have any update on this?

**Comment by giraffate** [comment]
ping from triage @Valentine-Mario. According to the triage procedure, I'm closing this because 2 weeks have passed with no activity. If you have more time to work on this, feel free to reopen this.

**Pull request closed without merging**